File: src/eqn/ast/EqnAst.java
package eqn.ast;

import eqn.parser.exception.EqnException;
import org.jetbrains.annotations.NotNull;

public class EqnAst {
    public EqnAstNode root;

    public EqnAst(@NotNull EqnAstNode root) {
        this.root = root;
    }

    public void simplify() throws EqnException {
        root = root.simplify();
        if (root.type == EqnAstNode.Type.Constant) {
            root = new EqnAstNodeDouble(Double.parseDouble(root.value));
        }
    }

    @Override
    public String toString() {
        return "EquationAst{" +
                "root=" + root +
                '}';
    }

    public double evaluate() throws EqnException {
        return root.evaluate();
    }
}

File: src/eqn/ast/EqnAst.kt
package eqn.ast

import eqn.ast.EqnAstNode
import eqn.parser.exception.EqnException

class EqnAst(root: EqnAstNode) {
    private var root: EqnAstNode?

    @Throws(EqnException::class)
    fun simplify() {
        root = root!!.simplify()
        if (root!!.type == EqnAstNode.Type.Constant) {
            root = EqnAstNodeDouble(root!!.value.toDouble())
        }
    }

    override fun toString(): String {
        return "EquationAst{" +
                "root=" + root +
                '}'
    }

    @Throws(EqnException::class)
    fun evaluate(): Double {
        return root!!.evaluate()
    }

    init {
        this.root = root
    }
}